### DISCUSSION TITLE
Go API support

### DISCUSSION THREAD

**Discussion by adonespitogo** [a-h/templ#979] [Ideas]
I'd like to programatically use this library in my project using Go API similar to [esbuild Go API](https://esbuild.github.io/api/) . Right now, I need to spawn a child process that executes `templ generate` to generate the go files. It would be nice to be able to use this library using pure go without the overhead of spawing a seprate process.

**Comment by a-h** [comment]
I can't really guarantee the API surface will remain stable but you can use the APIs that exist today.

Take a look at the `Generate` function in https://github.com/a-h/templ/blob/main/generator/generator.go and the `Parse` and `ParseString` functions in https://github.com/a-h/templ/blob/main/parser/v2/templatefile.go

That's all you need to use templ programatically. It's what templ uses internally itself: https://github.com/a-h/templ/blob/079e286e437b721ac5755cbe4aaab554b31a5456/cmd/templ/generatecmd/eventhandler.go#L220

**Reply by adonespitogo** [reply]
Thanks. This will do for now. 

I just have a question regarding the generated go files. Is it guarateed that the generated go files today be compatible with future releases of templ? Say I want to create a UI library with `templ components` and publish it on github, will it be compatible with the new templ-generated go files that uses newer `templ` version?

**Reply by a-h** [reply]
That's the goal. We're maintaining the runtime api surface, and don't intend to break that API compatibility, Go compatibility promise style.
